Fix exception handling in the prestub worker

janvorli opened the pull request:
There is a bug in the new exception handling when `ThePreStub` is called
from `CallDescrWorkerInternal` and the exception is propagated through
that. One of such cases is when a managed class is being initialized
during JITting and the constructor is in an assembly that's not found.
The bug results in skipping all the native frames upto a managed frame
that called that native chain that lead to the exception.
The effect is that some lock holders are skipped and the exception is
not converted to `TypeInitializationException`.

The fix is to ensure that when `ThePreStub` is called by
`CallDescrWorkerInternal`, the exception is not caught in the
`PreStubWorker`. It is left flowing into the native calling code instead.

On Windows, we also need to prevent the `ProcessCLRException` invocation
to call into the managed exception handling code for the case when the 
`ProcessCLRException` personality routine was on one of our asm helpers.

Close https://github.com/dotnet/runtime/issues/112502

janvorli committed:
Fix exception handling in prestub worker

There is a bug in the new exception handling when ThePreStub is called
from CallDescrWorkerInternal and the exception is propagated through
that. One of such cases is when a managed class is being initialized
during JITting and the constructor is in an assembly that's not found.
The bug results in skipping all the native frames upto a managed frame
that called that native chain that lead to the exception. In the
specific case I've mentioned, a lock in the native code is left in
locked state. That later leads to a hang. This was case was observed
with Roslyn invoking an analyzer where one of the dependencies was
missing.

The fix is to ensure that when ThePreStub is called by
CallDescrWorkerInternal, the exception is not caught in the
PreStubWorker. It is left flowing into the native calling code instead.

On Windows, we also need to prevent the ProcessCLRException invocation
to call into the managed exception handling code.

janvorli committed:
Regression test

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2556,17 +2566,33 @@ static PCODE PreStubWorker_Preemptive(
-    INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+    EX_TRY
+    {
+        bool propagateExceptionToNativeCode = IsCallDescrWorkerInternalReturnAddress(pTransitionBlock->m_ReturnAddress);
```
`CallDescrWorker` is guaranteed to run in cooperative mode, but `PreStubWorker_Preemptive` is preemptive entrypoint so it should not ever be called by called by CallDescrWorker. Are the changes in `PreStubWorker_Preemptive` reachable?

jkotas commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -2618,60 +2644,76 @@ extern "C" PCODE STDCALL PreStubWorker(TransitionBlock* pTransitionBlock, Method
-        INSTALL_UNWIND_AND_CONTINUE_HANDLER;
+        EX_TRY
+        {
+            bool propagateExceptionToNativeCode = IsCallDescrWorkerInternalReturnAddress(pTransitionBlock->m_ReturnAddress);
```
Does this problem exist in other places that use `INSTALL_MANAGED_EXCEPTION_DISPATCHER_EX

For example, can ExternalMethodFixupWorker or VSD_ResolveWorker end up appear to be called from CallDescrWorker?

janvorli replied:
That's a good point. I've checked and verified that both of the cases you've mentioned can be called via `CallDescrWorker`, so both suffer from the same issue in case there is an exception stemming from the resolution of the target. I think there are no other cases than the ones you've mentioned.

jkoritzinsky commented on src/tests/Regressions/coreclr/GitHub_76531/test76531.cs:
```diff
@@ -0,0 +1,43 @@
+        [Fact]
+        public static void TestEntryPoint()
+        {
+            File.Delete(Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), "dependencytodelete.dll"));
```
Instead of deleting it, could we instead change the csproj to set `Private="false"` on the ProjectReference so it isn't copied to the output in the first place?

janvorli replied:
Thanks for suggesting it, I'll give it a try.

janvorli committed:
Fix ExternalMethodFixupWorker and VSD_ResolveWorker too

These two need the same treatment when they end up being called from
CallDescrWorkerInternal.

Add regression tests for those cases too. These two cases don't result
in a visible failure, so the regression tests just exercise the specific
code paths.

janvorli committed:
Disable the regression test for NativeAOT / MonoAOT

janvorli commented:
/azp run runtime-coreclr crossgen2 outerloop

janvorli committed:
Reverting change in PrestubWorker_Preemptive

This change is not needed, as that function cannot be called via
CallDescrWorker.

janvorli commented:
@jkotas can you please take another look? I believe I have addressed all your feedback.

jkotas commented:
https://github.com/dotnet/runtime/pull/111937#discussion_r1933038306 does not appear addressed.

jkotas commented on src/tests/Regressions/coreclr/GitHub_76531/tailcallinvoker.il:
```diff
@@ -0,0 +1,17 @@
+
+.assembly extern legacy library mscorlib {}
+.assembly extern dependencytodelete {}
+.assembly 'lowlevel' {}
```
```suggestion
.assembly 'tailcallinvoker' {}
```
Should the assembly name match the file name?

janvorli replied:
Thank you for spotting this, I've forgotten to rename it after renaming the file itself locally.

jkotas commented on src/tests/Regressions/coreclr/GitHub_76531/test76531.csproj:
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <!-- Needed for mechanical merging of all remaining tests, this particular project may not actually need process isolation -->
```
Does the test need the process isolation? If yes, this comment should be deleted. If no, both this comment and the property should be deleted.

janvorli replied:
I need to check that.

janvorli committed:
Few forgotten cleanups

janvorli committed:
Remove the RequiresProcessIsolation

jkotas reviewed (approved):
LGTM!

janvorli committed:
Disable the regression test for WASM

janvorli committed:
Attempt to properly disable the test for WASM

janvorli committed:
Another attempt to prevent running the test on WASM
